### ISSUE TITLE
Example 02_transformers_tagger_bert.ipynb is broken.

### ISSUE THREAD

**Issue by lkhphuc** [explosion/thinc#313]
Hi, the examples number `02_transformers_tagger_bert.ipynb` is currently broken at `model.initialize(...)`.
Here are the stack trace:
```
---------------------------------------------------------------------------

Exception                                 Traceback (most recent call last)

<ipython-input-33-5f258b534bc1> in <module>()
      5 dev_Y = list(map(model.ops.asarray, dev_Y))  # convert to cupy if needed
      6 
----> 7 model.initialize(X=train_X[:5], Y=train_Y[:5])

5 frames

/usr/local/lib/python3.6/dist-packages/thinc/model.py in initialize(self, X, Y)
    277             validate_fwd_input_output(self.name, self._func, X, Y)
    278         if self._init is not None:
--> 279             self._init(self, X=X, Y=Y)
    280         return self
    281 

/usr/local/lib/python3.6/dist-packages/thinc/layers/chain.py in init(model, X, Y)
     78             layer.initialize(X=curr_input)
     79         if curr_input is not None:
---> 80             curr_input = layer.predict(curr_input)
     81     if model.layers[0].has_dim("nI"):
     82         model.set_dim("nI", model.layers[0].get_dim("nI"))

/usr/local/lib/python3.6/dist-packages/thinc/model.py in predict(self, X)
    293         only the output, instead of the `(output, callback)` tuple.
    294         """
--> 295         return self._func(self, X, is_train=False)[0]
    296 
    297     def finish_update(self, optimizer: Optimizer) -> None:

<ipython-input-12-3b8e8a6a55e1> in forward(model, texts, is_train)
     14             return_attention_masks=True,
     15             return_input_lengths=True,
---> 16             return_tensors="pt",
     17         )
     18         return TokensPlus(**token_data), lambda d_tokens: []

/usr/local/lib/python3.6/dist-packages/transformers/tokenization_utils.py in batch_encode_plus(self, batch_text_or_text_pairs, add_special_tokens, max_length, stride, truncation_strategy, pad_to_max_length, return_tensors, return_token_type_ids, return_attention_mask, return_overflowing_tokens, return_special_tokens_mask, return_offsets_mapping, **kwargs)
   1725                 tokens = [tokens]
   1726             else:
-> 1727                 tokens = self._tokenizer.encode_batch(batch_text_or_text_pairs)
   1728 
   1729         # Convert encoding to dict

/usr/local/lib/python3.6/dist-packages/tokenizers/implementations/base_tokenizer.py in encode_batch(self, sequences)
    152             A list of Encoding
    153         """
--> 154         return self._tokenizer.encode_batch(sequences)
    155 
    156     def decode(self, ids: List[int], skip_special_tokens: Optional[bool] = True) -> str:

Exception: Input must be a list[str] or list[(str, str)]
```
I believe there was something change from the `tokenizer` from HuggingFace but I'm not so sure.

**Comment by honnibal** [comment]
Thanks for the report, I'll look into it

**Comment by statisticallyfit** [comment]
Dear @honnibal  and @lkhphuc, 

Thanks for looking into this issue. I also got errors but a different kind than @lkhphuc 's error. 

At the line ...
`model.initialize(X=train_X[:5], Y=train_Y[:5])`

... here is my error stack trace: 

```
AttributeError                            Traceback (most recent call last)
<ipython-input-11-5f258b534bc1> in <module>
      5 dev_Y = list(map(model.ops.asarray, dev_Y))  # convert to cupy if needed
      6 
----> 7 model.initialize(X=train_X[:5], Y=train_Y[:5])

/development/bin/python/conda3_ana/envs/pythinc_env/lib/python3.6/site-packages/thinc/model.py in initialize(self, X, Y)
    277             validate_fwd_input_output(self.name, self._func, X, Y)
    278         if self._init is not None:
--> 279             self._init(self, X=X, Y=Y)
    280         return self
    281 

/development/bin/python/conda3_ana/envs/pythinc_env/lib/python3.6/site-packages/thinc/layers/chain.py in init(model, X, Y)
     78             layer.initialize(X=curr_input)
     79         if curr_input is not None:
---> 80             curr_input = layer.predict(curr_input)
     81     if model.layers[0].has_dim("nI"):
     82         model.set_dim("nI", model.layers[0].get_dim("nI"))

/development/bin/python/conda3_ana/envs/pythinc_env/lib/python3.6/site-packages/thinc/model.py in predict(self, X)
    293         only the output, instead of the `(output, callback)` tuple.
    294         """
--> 295         return self._func(self, X, is_train=False)[0]
    296 
    297     def finish_update(self, optimizer: Optimizer) -> None:

<ipython-input-5-fc76c8adeabc> in forward(model, texts, is_train)
      7     def forward(model, texts: List[List[str]], is_train: bool):
      8         tokenizer = model.attrs["tokenizer"]
----> 9         token_data = tokenizer.batch_encode_plus(
     10             [(text, None) for text in texts],
     11             add_special_tokens=True,

AttributeError: 'BertTokenizer' object has no attribute 'batch_encode_plus'
```

At runtime, `tokenizer` is of the `BertTokenizer` type (because of the CONFIG string specification)

[The `batch_encode_plus` is just in the general `PreTrainedTokenizer` class. ](https://huggingface.co/transformers/main_classes/tokenizer.html#transformers.PreTrainedTokenizer.batch_encode_plus). Doesn't seem to be there for `BertTokenizer`.

**Comment by lkhphuc** [comment]
Hi @statisticallyfit , I don't get your error, but I managed to get the notebook 02 running on Colab. Here are what I have to change:

* `return_input_lengths` of Tokenizer is not reliable so I pass in my own length.  And also add padding option.
```
@thinc.registry.layers("transformers_tokenizer.v1")
def TransformersTokenizer(name: str) -> Model[List[List[str]], TokensPlus]:
    def forward(model, texts: List[List[str]], is_train: bool):
        tokenizer = model.attrs["tokenizer"]
        token_data = tokenizer.batch_encode_plus(
            [(text, None) for text in texts],
            add_special_tokens=True,
            return_token_type_ids=True,
            return_attention_masks=True,
            return_input_lengths=True,
            pad_to_max_length=True,
            return_tensors="pt",
        )
        token_data['input_len'] = [len(text) for text in texts]
        return TokensPlus(**token_data), lambda d_tokens: []

    return Model("tokenizer", forward, attrs={"tokenizer": AutoTokenizer.from_pretrained(name)})
```

* Don't remove the BOS, EOS tokens, I think they are already removed. However code for "Restore BOS and EOS..." in the backward pass must still be included. I don't understand what's happening and what's broken here yet.
```
def convert_transformer_outputs(model, inputs_outputs, is_train):
    layer_inputs, torch_outputs = inputs_outputs
    torch_tokvecs: torch.Tensor = torch_outputs[0]
    torch_outputs = None  # free the memory as soon as we can
    lengths = list(layer_inputs.input_len)
    tokvecs: List[Array2d] = model.ops.unpad(torch2xp(torch_tokvecs), lengths)
#### Comment this line
    # tokvecs = [arr[1:-1] for arr in tokvecs]  # remove the BOS and EOS markers

    def backprop(d_tokvecs: List[Array2d]) -> ArgsKwargs:
        # Restore entries for BOS and EOS markers KEEP THESE FOLLOWING LINES
        row = model.ops.alloc2f(1, d_tokvecs[0].shape[1])
        d_tokvecs = [model.ops.xp.vstack((row, arr, row)) for arr in d_tokvecs]
        return ArgsKwargs(
            args=(torch_tokvecs,),
            kwargs={"grad_tensors": xp2torch(model.ops.pad(d_tokvecs))},
        )

    return tokvecs, backprop
```

**Comment by GoooIce** [comment]
```
0%
0/112 [00:01<?, ?it/s]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-9af39f1c45a7> in <module>()
     10             inputs, truths = zip(*batch)
     11             guesses, backprop = model(inputs, is_train=True)
---> 12             backprop(calculate_loss.get_grad(guesses, truths))
     13         model.finish_update(optimizer)
     14         optimizer.step_schedules()

1 frames
/usr/local/lib/python3.6/dist-packages/thinc/loss.py in get_grad(self, guesses, truths)
     49         if guesses.shape != target.shape:  # pragma: no cover
     50             err = f"Cannot calculate CategoricalCrossentropy loss: mismatched shapes: {guesses.shape} vs {target.shape}."
---> 51             raise ValueError(err)
     52         if guesses.any() > 1 or guesses.any() < 0:  # pragma: no cover
     53             err = f"Cannot calculate CategoricalCrossentropy loss with guesses outside the [0,1] interval."

ValueError: Cannot calculate CategoricalCrossentropy loss: mismatched shapes: (71, 17) vs (70, 17).
```

> 嗨@statisticallyfit，我没有收到您的错误，但是我设法使笔记本02在Colab上运行。这是我必须更改的内容：
> 
> * `return_input_lengths`Tokenizer的功能不可靠，所以我用了我自己的篇幅。并添加填充选项。
> 
> ```
> @thinc.registry.layers("transformers_tokenizer.v1")
> def TransformersTokenizer(name: str) -> Model[List[List[str]], TokensPlus]:
>     def forward(model, texts: List[List[str]], is_train: bool):
>         tokenizer = model.attrs["tokenizer"]
>         token_data = tokenizer.batch_encode_plus(
>             [(text, None) for text in texts],
>             add_special_tokens=True,
>             return_token_type_ids=True,
>             return_attention_masks=True,
>             return_input_lengths=True,
>             pad_to_max_length=True,
>             return_tensors="pt",
>         )
>         token_data['input_len'] = [len(text) for text in texts]
>         return TokensPlus(**token_data), lambda d_tokens: []
> 
>     return Model("tokenizer", forward, attrs={"tokenizer": AutoTokenizer.from_pretrained(name)})
> ```
> 
> * 不要删除BOS，EOS令牌，我认为它们已被删除。但是，向后传递中的“ Restore BOS and EOS ...”代码仍必须包括在内。我不明白发生了什么，这里还发生了什么。
> 
> ```
> def convert_transformer_outputs(model, inputs_outputs, is_train):
>     layer_inputs, torch_outputs = inputs_outputs
>     torch_tokvecs: torch.Tensor = torch_outputs[0]
>     torch_outputs = None  # free the memory as soon as we can
>     lengths = list(layer_inputs.input_len)
>     tokvecs: List[Array2d] = model.ops.unpad(torch2xp(torch_tokvecs), lengths)
> #### Comment this line
>     # tokvecs = [arr[1:-1] for arr in tokvecs]  # remove the BOS and EOS markers
> 
>     def backprop(d_tokvecs: List[Array2d]) -> ArgsKwargs:
> ##### DONT Comment these lines
>         # Restore entries for BOS and EOS markers
>         row = model.ops.alloc2f(1, d_tokvecs[0].shape[1])
>         d_tokvecs = [model.ops.xp.vstack((row, arr, row)) for arr in d_tokvecs]
>         return ArgsKwargs(
>             args=(torch_tokvecs,),
>             kwargs={"grad_tensors": xp2torch(model.ops.pad(d_tokvecs))},
>         )
> 
>     return tokvecs, backprop
> ```

@lkhphuc same code with you  0.0

**Comment by ramanova** [comment]
Hi @lkhphuc , Still getting
`---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-58-9af39f1c45a7> in <module>()
     10             inputs, truths = zip(*batch)
     11             guesses, backprop = model(inputs, is_train=True)
---> 12             backprop(calculate_loss.get_grad(guesses, truths))
     13         model.finish_update(optimizer)
     14         optimizer.step_schedules()

1 frames
/usr/local/lib/python3.6/dist-packages/thinc/loss.py in get_grad(self, guesses, truths)
     49         if guesses.shape != target.shape:  # pragma: no cover
     50             err = f"Cannot calculate CategoricalCrossentropy loss: mismatched shapes: {guesses.shape} vs {target.shape}."
---> 51             raise ValueError(err)
     52         if guesses.any() > 1 or guesses.any() < 0:  # pragma: no cover
     53             err = f"Cannot calculate CategoricalCrossentropy loss with guesses outside the [0,1] interval."

ValueError: Cannot calculate CategoricalCrossentropy loss: mismatched shapes: (71, 17) vs (70, 17).`

**Comment by honnibal** [comment]
Sorry for letting this slip all :(. Thanks for the helpful notes. The problem is I didn't pin to the transformers version, so we're getting incompatibilities as their library changes.

I think this is the right summary of actions:

* We need to set `pad_to_max_length=True` as @lkhphuc suggests
* We need to set `return_lengths=True` (changing from `return_input_lengths`)
* We need to change the field in `TokensPlus` to `length`, as that's what the value is called now.

I think this works. @lkhphuc the confusion in your code is I think that you're setting the input lengths to the *text* lengths. We need the number of tokens in the outputs so that we can unpad correctly.

**Comment by honnibal** [comment]
Looking at this again, I'm not 100% sure how the original code was supposed to work. I think there was actually a mistake and it was acting as though each token in the corpus would receive exactly one wordpiece? This obviously won't produce decent results. We need to calculate some sort of alignment.

We can use @tamuhey's tokenizations library to get the alignment, that's no problem. But I want a more elegant way to actually apply it...

But even after years of numpy I still get incredibly frustrated trying to get the right magic for indexing. Honestly it's the unhappiest I ever am while programming, I wish I could just write the stupid loop in Cython. The logic is always completely trivial. Here's what we want:

```python

def fill_aligned(dest: Floats3d, src: Floats3d, indices: List[List[List[int]]]:
    for i, doc_aligns in enumerate(indices):
        for dest_j, token_aligns in enumerate(doc_aligns):
            for src_j in token_aligns:
                dest[i, dest_j] = src[i, src_j]
```


Anyway, here's where I got to. I figured if I calculated a 2d mask there would be some way to set the rows in one go. But I can't seem to figure out how to do it (╯°□°)╯︵ ┻━┻

```python
import tokenizations

def get_alignment_mask(orig_tokens, tokenizer, input_ids):
    """Create a boolean mask indicating whether the wordpiece ends a token."""
    wp_tokens = [tokenizer.convert_ids_to_tokens(row) for row in input_ids]
    mask = numpy.zeros(input_ids.shape, dtype="b")
    for i, (orig, wp) in enumerate(zip(orig_tokens, wp_tokens)):
        a2b, b2a = tokenizations.get_alignments(orig, wp)
        for j, j_pieces in enumerate(a2b):
            if j_pieces:
                mask[i, j_pieces[-1]] = 1
    return mask
```

We want to take the 3d tensor produced by the transformer, and select a 3d subtensor according to the mask. We then want to split it using the lengths of the original inputs, so that we get a `List[Floats2d]` that matches up to the labels. Then we need to do the opposite in the backward pass: we need to take the `List[Floats2d]` of the gradients, pad it, and then use the mask to create an expanded tensor that matches the transformer output.
